### ISSUE TITLE
Blank Inference v0.5.0 when using DeepSpeech binary

### ISSUE THREAD

**Issue by AASHISHAG** [mozilla/DeepSpeech#2397]
- **OS Platform and Distribution:**  Linux Ubuntu 16.04
- **TensorFlow installed from (our builds, or upstream TensorFlow)**: pip3 install tensorflow==1.13.1
- **TensorFlow version (use command below)**: 'v1.13.1-0-g6612da8951' 1.13.1
- **Python version**: Python 3.5.2
- **Bazel version (if compiling from source)**: Build label: 0.19.2
- **GCC version (if compiling from source)**: gcc version 5.4.0 20160609
- **CUDA/cuDNN version**: CUDA Version 10.0.130
- **GPU model and memory**: Nvidia Quadro RTX 6000 (3*24GB)

I am trying to use deepspeech python binary (pip3 install deepspeech==0.5.0) to test my german deepspeech model (0.5.0). But surprisingly I am getting blank inference results. Below is the log:

```
(env-run-deepspeech) agarwal@LTLab.lan@wika:~$ deepspeech --model models/model_mozilla.pb --alphabet dataset-german/data/alphabet.txt --lm dataset-german/lm.binary --trie dataset-german/trie --audio dataset-german/de_3.wav
Loading model from file models/model_mozilla.pb
TensorFlow: v1.13.1-10-g3e0cc53
DeepSpeech: v0.5.0-0-g3db7a99
Warning: reading entire model file into memory. Transform model file into an mmapped graph to reduce heap usage.
2019-10-02 21:24:20.350300: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 AVX512F FMA
2019-10-02 21:24:20.513925: E tensorflow/core/framework/op_kernel.cc:1325] OpKernel ('op: "UnwrapDatasetVariant" device_type: "CPU"') for unknown op: UnwrapDatasetVariant
2019-10-02 21:24:20.513964: E tensorflow/core/framework/op_kernel.cc:1325] OpKernel ('op: "WrapDatasetVariant" device_type: "GPU" host_memory_arg: "input_handle" host_memory_arg: "output_handle"') for unknown op: WrapDatasetVariant
2019-10-02 21:24:20.513974: E tensorflow/core/framework/op_kernel.cc:1325] OpKernel ('op: "WrapDatasetVariant" device_type: "CPU"') for unknown op: WrapDatasetVariant
2019-10-02 21:24:20.514124: E tensorflow/core/framework/op_kernel.cc:1325] OpKernel ('op: "UnwrapDatasetVariant" device_type: "GPU" host_memory_arg: "input_handle" host_memory_arg: "output_handle"') for unknown op: UnwrapDatasetVariant
Loaded model in 0.168s.
Loading language model from files dataset-german/lm.binary dataset-german/trie
Loaded language model in 0.45s.
Running inference.

Inference took 4.508s for 3.264s audio file.
```

Here are the training details. I created my own lm.binary and trie to train the model, using the below commands.

Language Model: KenLM
```
$kenlm/build/bin/lmplz --text $export_path/clean_vocab.txt --arpa $export_path/words.arpa --o 3
$kenlm/build/bin/build_binary -T -s $export_path/words.arpa $export_path/lm.binary
```
Native Client
```
$ git clone https://github.com/mozilla/tensorflow.git
$ git checkout origin/r1.14
$ cd tensorflow
$ ./configure
$ cd tensorflow
$ ln -s ../DeepSpeech/native_client ./
rkspace_status_command="bash native_client/bazel_workspace_status_cmd.sh" --config=monolithic -c opt --copt=-O3 --copt="-D_GLIBCXX_USE_CXX11_ABI=0" --copt=-fvisibility=hidden //native_client:libdeepspeech.so //native_client:generate_trie --config=cuda
$ TFDIR=~/tensorflow
$ cd ../DeepSpeech/native_client
$ make deepspeech
```
TRIE
```
$ DeepSpeech/native_client/generate_trie $export_path/alphabet.txt $export_path/lm.binary $export_path/clean_vocab.txt $export_path/trie
```

Training
```
./DeepSpeech.py --train_files ../german-speech-corpus/data_tuda+voxforge/tmp3/train_100.csv --dev_files ../german-speech-corpus/data_tuda+voxforge/tmp3/dev.csv --test_files ../german-speech-corpus/data_tuda+voxforge/tmp3/test.csv --alphabet_config_path ../deepspeech-german/data/alphabet.txt --lm_trie_path ../dataset-german/trie --lm_binary_path ../dataset-german/lm.binary --test_batch_size 36 --train_batch_size 24 --dev_batch_size 36 --epochs 75 --learning_rate 0.0001 --dropout_rate 0.30 --export_dir ../models
```
I have created a repository with all training details, please refer them here:
```
https://github.com/AASHISHAG/deepspeech-german
```
Kindly let me know if you need any other specific information.

**Comment by lissyx** [comment]
> I trained on German language, but surprisingly I am getting blank inference results.

I tried to help you, but surprisingly, since there is no information on how you performed your training, I am getting blank results.

**Comment by AASHISHAG** [comment]
@lissyx : I have updated the above comment. Please check.

**Comment by lissyx** [comment]
@AASHISHAG Why are you rebuilding the `deepspeech` inference code ?

Training parameters without training log is not useful. How much data do you have ? What WER did you got ? How much epoch did complete ?

**Comment by AASHISHAG** [comment]
@lissyx : I was trying to follow the ReadMe for creating TRIE and I mistakenly followed an extra step, Please ignore it.

I am attaching the training logs for a model trained on ~302 hours of data, which produces WER of 24.48%. It ran for 12 Epochs.

```
./DeepSpeech.py --train_files ../german-speech-corpus/data_tuda+voxforge+mozilla/train_100.csv --dev_files ../german-speech-corpus/data_tuda+voxforge+mozilla/dev.csv --test_files ../german-speech-corpus/data_tuda+voxforge+mozilla/test.csv --alphabet_config_path ../deepspeech-german/data/alphabet.txt --lm_trie_path ../dataset-german/trie --lm_binary_path ../dataset-german/lm.binary --test_batch_size 36 --train_batch_size 24 --dev_batch_size 36 --epochs 75 --learning_rate 0.0001 --dropout_rate 0.30 --export_dir ../models

WARNING:tensorflow:From /home/LTLab.lan/agarwal/python-environments/env/lib/python3.5/site-packages/tensorflow/python/data/ops/dataset_ops.py:429: py_func (from tensorflow.python.ops.script_ops) is deprecated and will be removed in a future version.
Instructions for updating:
tf.py_func is deprecated in TF V2. Instead, use
    tf.py_function, which takes a python function which manipulates tf eager
    tensors instead of numpy arrays. It's easy to convert a tf eager tensor to
    an ndarray (just call tensor.numpy()) but having access to eager tensors
    means `tf.py_function`s can use accelerators such as GPUs as well as
    being differentiable using a gradient tape.

WARNING:tensorflow:From /home/LTLab.lan/agarwal/python-environments/env/lib/python3.5/site-packages/tensorflow/python/data/ops/iterator_ops.py:358: colocate_with (from tensorflow.python.framework.ops) is deprecated and will be removed in a future version.
Instructions for updating:
Colocations handled automatically by placer.
WARNING:tensorflow:From /home/LTLab.lan/agarwal/python-environments/env/lib/python3.5/site-packages/tensorflow/contrib/rnn/python/ops/lstm_ops.py:696: to_int64 (from tensorflow.python.ops.math_ops) is deprecated and will be removed in a future version.
Instructions for updating:
Use tf.cast instead.
I Initializing variables...
I STARTING Optimization
Epoch 0 |   Training | Elapsed Time: 0:04:24 | Steps: 445 | Loss: 176.860081
Epoch 0 | Validation | Elapsed Time: 0:00:14 | Steps: 63 | Loss: 155.113629 | Dataset: ../german-speech-corpus/data_tuda+voxforge+mozilla/dev.csv
I Saved new best validating model with loss 155.113629 to: /home/LTLab.lan/agarwal/.local/share/deepspeech/checkpoints/best_dev-445
Epoch 1 |   Training | Elapsed Time: 0:04:12 | Steps: 445 | Loss: 118.930347
Epoch 1 | Validation | Elapsed Time: 0:00:14 | Steps: 63 | Loss: 122.372937 | Dataset: ../german-speech-corpus/data_tuda+voxforge+mozilla/dev.csv
WARNING:tensorflow:From /home/LTLab.lan/agarwal/python-environments/env/lib/python3.5/site-packages/tensorflow/python/training/saver.py:966: remove_checkpoint (from tensorflow.python.training.checkpoint_management) is deprecated and will be removed in a future version.
Instructions for updating:
Use standard file APIs to delete files with this prefix.
I Saved new best validating model with loss 122.372937 to: /home/LTLab.lan/agarwal/.local/share/deepspeech/checkpoints/best_dev-890
Epoch 2 |   Training | Elapsed Time: 0:04:11 | Steps: 445 | Loss: 98.190498
Epoch 2 | Validation | Elapsed Time: 0:00:14 | Steps: 63 | Loss: 107.649564 | Dataset: ../german-speech-corpus/data_tuda+voxforge+mozilla/dev.csv
I Saved new best validating model with loss 107.649564 to: /home/LTLab.lan/agarwal/.local/share/deepspeech/checkpoints/best_dev-1335
Epoch 3 |   Training | Elapsed Time: 0:04:12 | Steps: 445 | Loss: 85.011518
Epoch 3 | Validation | Elapsed Time: 0:00:14 | Steps: 63 | Loss: 91.951983 | Dataset: ../german-speech-corpus/data_tuda+voxforge+mozilla/dev.csv
I Saved new best validating model with loss 91.951983 to: /home/LTLab.lan/agarwal/.local/share/deepspeech/checkpoints/best_dev-1780
Epoch 4 |   Training | Elapsed Time: 0:04:11 | Steps: 445 | Loss: 75.027806
Epoch 4 | Validation | Elapsed Time: 0:00:14 | Steps: 63 | Loss: 86.279818 | Dataset: ../german-speech-corpus/data_tuda+voxforge+mozilla/dev.csv
I Saved new best validating model with loss 86.279818 to: /home/LTLab.lan/agarwal/.local/share/deepspeech/checkpoints/best_dev-2225
Epoch 5 |   Training | Elapsed Time: 0:04:10 | Steps: 445 | Loss: 67.094796
Epoch 5 | Validation | Elapsed Time: 0:00:14 | Steps: 63 | Loss: 85.204753 | Dataset: ../german-speech-corpus/data_tuda+voxforge+mozilla/dev.csv
I Saved new best validating model with loss 85.204753 to: /home/LTLab.lan/agarwal/.local/share/deepspeech/checkpoints/best_dev-2670
Epoch 6 |   Training | Elapsed Time: 0:04:11 | Steps: 445 | Loss: 60.479755
Epoch 6 | Validation | Elapsed Time: 0:00:14 | Steps: 63 | Loss: 81.458550 | Dataset: ../german-speech-corpus/data_tuda+voxforge+mozilla/dev.csv
I Saved new best validating model with loss 81.458550 to: /home/LTLab.lan/agarwal/.local/share/deepspeech/checkpoints/best_dev-3115
Epoch 7 |   Training | Elapsed Time: 0:04:11 | Steps: 445 | Loss: 54.812813
Epoch 7 | Validation | Elapsed Time: 0:00:14 | Steps: 63 | Loss: 76.570380 | Dataset: ../german-speech-corpus/data_tuda+voxforge+mozilla/dev.csv
I Saved new best validating model with loss 76.570380 to: /home/LTLab.lan/agarwal/.local/share/deepspeech/checkpoints/best_dev-3560
Epoch 8 |   Training | Elapsed Time: 0:04:11 | Steps: 445 | Loss: 49.687799
Epoch 8 | Validation | Elapsed Time: 0:00:14 | Steps: 63 | Loss: 75.798708 | Dataset: ../german-speech-corpus/data_tuda+voxforge+mozilla/dev.csv
I Saved new best validating model with loss 75.798708 to: /home/LTLab.lan/agarwal/.local/share/deepspeech/checkpoints/best_dev-4005
Epoch 9 |   Training | Elapsed Time: 0:04:11 | Steps: 445 | Loss: 45.490483
Epoch 9 | Validation | Elapsed Time: 0:00:14 | Steps: 63 | Loss: 74.208162 | Dataset: ../german-speech-corpus/data_tuda+voxforge+mozilla/dev.csv
I Saved new best validating model with loss 74.208162 to: /home/LTLab.lan/agarwal/.local/share/deepspeech/checkpoints/best_dev-4450
Epoch 10 |   Training | Elapsed Time: 0:04:11 | Steps: 445 | Loss: 41.527083
Epoch 10 | Validation | Elapsed Time: 0:00:14 | Steps: 63 | Loss: 72.952417 | Dataset: ../german-speech-corpus/data_tuda+voxforge+mozilla/dev.csv
I Saved new best validating model with loss 72.952417 to: /home/LTLab.lan/agarwal/.local/share/deepspeech/checkpoints/best_dev-4895
Epoch 11 |   Training | Elapsed Time: 0:04:11 | Steps: 445 | Loss: 38.197429
Epoch 11 | Validation | Elapsed Time: 0:00:14 | Steps: 63 | Loss: 75.873461 | Dataset: ../german-speech-corpus/data_tuda+voxforge+mozilla/dev.csv
WARNING:tensorflow:From /home/LTLab.lan/agarwal/python-environments/env/lib/python3.5/site-packages/tensorflow/python/training/saver.py:1266: checkpoint_exists (from tensorflow.python.training.checkpoint_management) is deprecated and will be removed in a future version.
Instructions for updating:
Use standard file APIs to check for files with this prefix.
I Early stop triggered as (for last 4 steps) validation loss: 75.873461 with standard deviation: 1.164670 and mean: 74.319763
I FINISHED optimization in 0:53:36.295521
I Restored variables from best validation checkpoint at /home/LTLab.lan/agarwal/.local/share/deepspeech/checkpoints/best_dev-4895, step 4895
Testing model on ../german-speech-corpus/data_tuda+voxforge+mozilla/test.csv
Test epoch | Steps: 190 | Elapsed Time: 0:06:19
WARNING:tensorflow:From /home/LTLab.lan/agarwal/python-environments/env/lib/python3.5/site-packages/tensorflow/python/tools/freeze_graph.py:232: convert_variables_to_constants (from tensorflow.python.framework.graph_util_impl) is deprecated and will be removed in a future version.
Instructions for updating:
Use tf.compat.v1.graph_util.convert_variables_to_constants
WARNING:tensorflow:From /home/LTLab.lan/agarwal/python-environments/env/lib/python3.5/site-packages/tensorflow/python/framework/graph_util_impl.py:245: extract_sub_graph (from tensorflow.python.framework.graph_util_impl) is deprecated and will be removed in a future version.
Instructions for updating:
Use tf.compat.v1.graph_util.extract_sub_graph
Test on ../german-speech-corpus/data_tuda+voxforge+mozilla/test.csv - WER: 0.244884, CER: 0.106593, loss: 34.352589
--------------------------------------------------------------------------------
WER: 3.000000, CER: 0.857143, loss: 15.761712
 - src: "nochmal"
 - res: "noch war es"
--------------------------------------------------------------------------------
WER: 2.500000, CER: 0.400000, loss: 51.369308
 - src: "dialogstatus zurücksetzen"
 - res: "die noch status zu setzen"
--------------------------------------------------------------------------------
WER: 2.000000, CER: 0.142857, loss: 3.455172
 - src: "nochmal"
 - res: "noch mal"
--------------------------------------------------------------------------------
WER: 2.000000, CER: 0.142857, loss: 3.665473
 - src: "nochmal"
 - res: "noch mal"
--------------------------------------------------------------------------------
WER: 2.000000, CER: 0.071429, loss: 5.423339
 - src: "weiterschlafen"
 - res: "weiter schlafen"
--------------------------------------------------------------------------------
WER: 2.000000, CER: 0.071429, loss: 5.625791
 - src: "weiterschlafen"
 - res: "weiter schlafen"
--------------------------------------------------------------------------------
WER: 2.000000, CER: 0.071429, loss: 6.189450
 - src: "weiterschlafen"
 - res: "weiter schlafen"
--------------------------------------------------------------------------------
WER: 2.000000, CER: 0.142857, loss: 6.387710
 - src: "nochmal"
 - res: "noch mal"
--------------------------------------------------------------------------------
WER: 2.000000, CER: 0.142857, loss: 7.680886
 - src: "nochmal"
 - res: "noch mal"
--------------------------------------------------------------------------------
WER: 2.000000, CER: 0.625000, loss: 8.305445
 - src: "nächstes"
 - res: "ist es"
--------------------------------------------------------------------------------
I Exporting the model...
I Models exported at ../models
```

**Comment by lissyx** [comment]
> I am attaching the training logs for a model trained on ~302 hours of data, which produces WER of 24.48%. It ran for 12 Epochs.

Now this is getting interesting, because it does not seems that you have not trained enough. Can you reproduce that on latest master, please ?

**Comment by AASHISHAG** [comment]
@lissyx : Does it mean we have to have train over 75 epochs? 

As you have asked, I trained the same data on master. I also had to create a new TRIE on the master. Here are the logs: 
[nohup.txt](https://github.com/mozilla/DeepSpeech/files/3688254/nohup.txt)

Then I installed: 

`pip3 install deepspeech==0.6.0a8`

and ran:

```
(env-deepspeech-german-test) agarwal@LTLab.lan@wika:~/deepspeech-german-test$ deepspeech --model models/output_graph.pb --alphabet dependencies/alphabet.txt --lm dependencies/lm.binary --trie dependencies/trie --audio ../dataset-german/de_1.wav
Loading model from file models/output_graph.pb
TensorFlow: v1.14.0-16-g3b4ce37
DeepSpeech: v0.6.0-alpha.8-0-gf0e9541
Warning: reading entire model file into memory. Transform model file into an mmapped graph to reduce heap usage.
2019-10-03 23:55:38.338905: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 AVX512F FMA
Loaded model in 0.166s.
Loading language model from files dependencies/lm.binary dependencies/trie
Loaded language model in 0.000183s.
Running inference.
auf meinen in diesem jahr mit
Inference took 2.562s for 1.490s audio file.
```

Unlike with v0.5.0, the master didn't produce blank inference.

Could you please guide, how can I get it fix in v0.5.0.

**Comment by lissyx** [comment]
> Could you please guide, how can I get it fix in v0.5.0?

First, you should use `v0.5.1`. Second, it's inconsistent. Your model, if it was trained with `v0.5.1` would not work with `0.6.0a8`. So I don't know what you mixed together ...

**Comment by lissyx** [comment]
@AASHISHAG Could you update us on your status ?

**Comment by AASHISHAG** [comment]
@lissyx : As you had suggested, python binary v0.5.1 worked well with a model trained on v0.5.0. 

The python binary v0.5.0 seems inconsistent and gave blank inference with a model trained on v0.5.0.

Thanks for the direction. I am closing the ticket.

**Comment by lissyx** [comment]
> The python binary v0.5.0 seems inconsistent and gave blank inference with a model trained on v0.5.0.

Indeed, v0.5.0 was bugged and v0.5.1 was a bugfix release.
